File: InsertInterval.cc
#include <iostream>
#include <vector>
using namespace std;

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

vector<Interval> insert0(vector<Interval> &intervals, Interval newInterval) {
    vector<Interval> result;

    bool inserted = false;
    for (int i = 0; (!inserted || i < intervals.size()); )
    {
        Interval curInterval ;
        if (inserted || ((i < intervals.size()) && (intervals[i].start <= newInterval.start)))
            curInterval = intervals[i++];
        else {
            curInterval = newInterval;
            inserted = true;
        }

        if (!result.empty()) {
            Interval prev = result.back();
            if (prev.end >= curInterval.start) {
                result.pop_back();
                Interval newPrev(prev.start, max(prev.end, curInterval.end));
                result.push_back(newPrev);
            } else
                result.push_back(curInterval);
        } else
            result.push_back(curInterval);

    }


    return result;
}

//Take into account the assumption that intervals are non-overlapping
vector<Interval> insert1(vector<Interval> &intervals, Interval newInterval) {
    vector<Interval> result;

    int i;
    for (i = 0; i < intervals.size(); ++i)
    {
        if ((intervals[i].end < newInterval.start))
            result.push_back(intervals[i]);
        else if ((intervals[i].start > newInterval.end))
            break;
        else {
            newInterval.start = min(intervals[i].start, newInterval.start);
            newInterval.end = max(intervals[i].end, newInterval.end);
        }
    }
    
    result.push_back(newInterval);

    for (; i < intervals.size(); ++i)
        result.push_back(intervals[i]);

    return result;
}

// More elegant logic
vector<Interval> insert(vector<Interval> &intervals, Interval newInterval) {
    
    vector<Interval> result;
    
    int i = 0;
    for (; (i < intervals.size()) && (intervals[i].start < newInterval.start); ++i)
    {
       result.push_back(intervals[i]);
    }
    
    if (result.empty() || (newInterval.start > result.back().end)) {
        result.push_back(newInterval);
    } else {
        result.back().end = max(result.back().end, newInterval.end);
    }
    
    for (; i < intervals.size(); ++i)
    {
        if (intervals[i].start > result.back().end)
            result.push_back(intervals[i]);
        else {
            result.back().end = max(result.back().end, intervals[i].end);
        }
    }
    
    return result;
}

int main(int argc, char** argv)
{
    vector<Interval> intervals = {
        Interval(1, 2),
        Interval(3, 5), 
        Interval(6, 7), 
        Interval(8, 10), 
        Interval(12, 16)
    };
    Interval newInterval(4, 9);
    vector<Interval> result = insert(intervals, newInterval);
    for (int i = 0; i < result.size(); ++i)
        cout << result[i].start << ", " << result[i].end << endl;

    return 0;
}

File: InsertInterval.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class InsertInterval {

	//O(n) time and O(n) space, non-inplace solution.
    public List<Interval> insert0(List<Interval> intervals, Interval newInterval) {
        List<Interval> result = new ArrayList<>();
        int i = 0;
        for (;i < intervals.size() && intervals.get(i).end < newInterval.start; result.add(intervals.get(i++)));
        Interval mergedInterval = new Interval(newInterval.start, newInterval.end);
        for(; i < intervals.size() && intervals.get(i).start <= mergedInterval.end; ++i) {
            mergedInterval.start = Math.min(mergedInterval.start, intervals.get(i).start);
            mergedInterval.end = Math.max(mergedInterval.end, intervals.get(i).end);
        }
        result.add(mergedInterval);
        for(; i < intervals.size(); result.add(intervals.get(i++)));
        return result;
    }

    //Same algorithm and performance as above, but use iterator instead of index to do iteration.
    //This is better than above because if the list is linked list, list.get(i) could be very expensive.
    public List<Interval> insert1(List<Interval> intervals, Interval newInterval) {
        List<Interval> result = new ArrayList<>();
        ListIterator<Interval> iter = intervals.listIterator();
        Interval mergedInterval = new Interval(newInterval.start, newInterval.end);
        boolean isInserted = false;
        for (;iter.hasNext();) {
            Interval curInterval = iter.next();
            if (curInterval.end < newInterval.start) {
               result.add(curInterval);
            } else if (curInterval.start <= newInterval.end) {
               mergedInterval.start = Math.min(mergedInterval.start, curInterval.start);
               mergedInterval.end = Math.max(mergedInterval.end, curInterval.end);                       
            } else {
               if (!isInserted) {
                   result.add(mergedInterval);
                   isInserted = true;
               }
               result.add(curInterval);
            }
        }
        if (!isInserted) {
            result.add(mergedInterval);
        }
        return result;
    }

    //In place solution. O(n^2) time and O(1) space. Same algorithm as above.
    public List<Interval> insert(List<Interval> intervals, Interval newInterval) {
        ListIterator<Interval> iter = intervals.listIterator();
        Interval mergedInterval = new Interval(newInterval.start, newInterval.end);
        boolean isInserted = false;
        for (;iter.hasNext();) {
               Interval curInterval = iter.next();
               if (curInterval.end >= newInterval.start) {
                   if (curInterval.start <= newInterval.end) {
                        mergedInterval.start = Math.min(mergedInterval.start, curInterval.start);
                        mergedInterval.end = Math.max(mergedInterval.end, curInterval.end);
                        iter.remove();
                   } else {
                       if (!isInserted) {
                           iter.previous();
                           iter.add(mergedInterval);
                           isInserted = true;
                       }
                   }
               }
        }
        if (!isInserted) {
            iter.add(mergedInterval);
        }
        return intervals;
    }
}